### PR TITLE
[OAP-1498][oap-native-sql] adding stddev_samp support

### PR DESCRIPTION
**Pull request by rui-mo**

### CODE REVIEW

**Comment by zhouyuan** [comment]
#1498

**Comment by rui-mo** [comment]
#### Query test result
#####row:
scala> time{spark.sql("select ss_item_sk, stddev_samp(ss_quantity) as store_sales_quantitystdev, stddev_samp(ss_list_price), stddev_samp(ss_ticket_number) from store_sales group by ss_item_sk order by ss_item_sk  limit 100").show}
+----------+-------------------------+--------------------------+---------------------------------------------+
|ss_item_sk|store_sales_quantitystdev|stddev_samp(ss_list_price)|stddev_samp(CAST(ss_ticket_number AS DOUBLE))|
+----------+-------------------------+--------------------------+---------------------------------------------+
|         1|       29.098844757713387|        45.684987487501004|                           6954564.2392293895|
|         2|       28.913715979808796|         45.67895110737274|                            6878413.014046009|
|         3|       29.245580021306424|          45.5064186915396|                            6982634.488898209|
|         4|        28.78873367452031|         46.75743482223979|                            6961028.661100444|
|         5|        29.10522946982835|        46.836166737158784|                            7075129.993478093|
|         6|       28.423247205783728|         47.15163292278171|                            7077872.927363345|
|         7|       28.986213962525092|        46.021991567234764|                            6891546.030064147|
|         8|        29.18407634779071|         46.84541619046964|                            6879552.527811598|
|         9|       28.828001208893188|        45.775777897290624|                            6960318.322029455|
|        10|       28.708234657077814|         45.16074014077233|                            6935612.930091468|
|        11|       28.669644445193697|         44.18214018142708|                            6946935.331484145|
|        12|       29.822617768430128|         45.00871667295131|                            7003550.311610291|
|        13|       28.521287693023258|         46.14608806648435|                            6917378.062627748|
|        14|       29.014408610915886|         46.36279201815749|                            6980381.787024499|
|        15|        28.42171789412903|         46.74001626767024|                            6818545.977606066|
|        16|       28.630078175297328|           45.937345156158|                            6998882.089007459|
|        17|       29.554911275819062|        46.308432282054504|                            6954240.964916541|
|        18|       28.762456466837115|         45.44823887552419|                            7114689.786164044|
|        19|       28.872599229536743|        45.913689187625245|                            7093953.951919404|
|        20|       29.376501344444247|         45.72323578704782|                            6878579.662488168|
+----------+-------------------------+--------------------------+---------------------------------------------+
only showing top 20 rows

Elapsed time: 17556ms


scala> time{spark.sql("select count(ss_ext_tax), stddev_samp(ss_ext_tax), stddev_samp(ss_item_sk), stddev_samp(ss_sales_price), stddev_samp(ss_addr_sk) from store_sales where limit 100").show}
+-----------------+-----------------------+---------------------------------------+---------------------------+---------------------------------------+
|count(ss_ext_tax)|stddev_samp(ss_ext_tax)|stddev_samp(CAST(ss_item_sk AS DOUBLE))|stddev_samp(ss_sales_price)|stddev_samp(CAST(ss_addr_sk AS DOUBLE))|
+-----------------+-----------------------+---------------------------------------+---------------------------+---------------------------------------+
|        275046633|     126.56286593257153|                      58892.23320053515|         34.630854050151605|                      288644.8411860637|
+-----------------+-----------------------+---------------------------------------+---------------------------+---------------------------------------+

Elapsed time: 7845ms

#####col:
groupby no codegen:
scala> time{spark.sql("select ss_item_sk, stddev_samp(ss_quantity) as store_sales_quantitystdev, stddev_samp(ss_list_price), stddev_samp(ss_ticket_number) from store_sales group by ss_item_sk order by ss_item_sk  limit 100").show}
+----------+-------------------------+--------------------------+---------------------------------------------+
|ss_item_sk|store_sales_quantitystdev|stddev_samp(ss_list_price)|stddev_samp(CAST(ss_ticket_number AS DOUBLE))|
+----------+-------------------------+--------------------------+---------------------------------------------+
|         1|       29.098844757713387|        45.684987487501026|                            6954564.239229389|
|         2|       28.913715979808796|         45.67895110737276|                             6878413.01404601|
|         3|       29.245580021306413|          45.5064186915396|                             6982634.48889821|
|         4|        28.78873367452032|        46.757434822239794|                            6961028.661100443|
|         5|        29.10522946982835|          46.8361667371588|                            7075129.993478092|
|         6|       28.423247205783735|        47.151632922781715|                            7077872.927363345|
|         7|        28.98621396252509|         46.02199156723476|                            6891546.030064147|
|         8|         29.1840763477907|         46.84541619046964|                            6879552.527811596|
|         9|        28.82800120889318|        45.775777897290624|                            6960318.322029454|
|        10|       28.708234657077796|         45.16074014077233|                            6935612.930091465|
|        11|       28.669644445193697|         44.18214018142707|                            6946935.331484145|
|        12|       29.822617768430124|        45.008716672951294|                             7003550.31161029|
|        13|        28.52128769302326|         46.14608806648437|                            6917378.062627748|
|        14|         29.0144086109159|         46.36279201815749|                            6980381.787024497|
|        15|       28.421717894129024|         46.74001626767024|                            6818545.977606066|
|        16|       28.630078175297324|         45.93734515615802|                           6998882.0890074605|
|        17|       29.554911275819055|         46.30843228205449|                            6954240.964916546|
|        18|        28.76245646683711|          45.4482388755242|                            7114689.786164045|
|        19|        28.87259922953674|        45.913689187625245|                            7093953.951919402|
|        20|        29.37650134444424|        45.723235787047805|                            6878579.662488163|
+----------+-------------------------+--------------------------+---------------------------------------------+
only showing top 20 rows

Elapsed time: 26994ms

groupby with codegen:

scala> time{spark.sql("select ss_item_sk, stddev_samp(ss_quantity) as store_sales_quantitystdev, stddev_samp(ss_list_price), stddev_samp(ss_ticket_number) from store_sales group by ss_item_sk order by ss_item_sk  limit 100").show}
+----------+-------------------------+--------------------------+---------------------------------------------+
|ss_item_sk|store_sales_quantitystdev|stddev_samp(ss_list_price)|stddev_samp(CAST(ss_ticket_number AS DOUBLE))|
+----------+-------------------------+--------------------------+---------------------------------------------+
|         1|       31.069967002932536|         48.41008394114372|                           6954564.2392293895|
|         2|        30.66174886272615|        47.681808381512276|                            6878413.014046009|
|         3|       30.442250662956965|         47.46347255493236|                            6982634.488898207|
|         4|         30.0407470036144|         48.58808088454524|                            6961028.661100446|
|         5|       30.552962301935743|         49.38439941736051|                            7075129.993478092|
|         6|        29.52804473996775|         48.42405043803398|                            7077872.927363348|
|         7|         31.0816842324423|         48.46162877881877|                            6891546.030064146|
|         8|        30.89443293725506|        49.629648585910175|                              6879552.5278116|
|         9|       30.089147228069034|        47.418590671398285|                            6960318.322029453|
|        10|       30.210749395258638|         48.17449581015934|                            6935612.930091466|
|        11|       29.795580619505337|         46.40456600293818|                           6946935.3314841455|
|        12|       30.859915634322064|         47.02131505037968|                            7003550.311610292|
|        13|        30.12285582637442|        48.368563541313954|                            6917378.062627747|
|        14|       30.789936112923954|         48.58091914984182|                              6980381.7870245|
|        15|       29.985446117097954|         48.80980124952307|                            6818545.977606066|
|        16|       30.174576065123194|         47.87510805267051|                            6998882.089007459|
|        17|       30.456853262265188|          48.1188986563067|                            6954240.964916545|
|        18|       29.946963599701363|         47.53902976938453|                            7114689.786164045|
|        19|       30.744165009957044|        48.267782260722264|                            7093953.951919405|
|        20|       31.002865951274533|        47.994500842237144|                            6878579.662488164|
+----------+-------------------------+--------------------------+---------------------------------------------+
only showing top 20 rows

Elapsed time: 27169ms

scala> time{spark.sql("select count(ss_ext_tax), stddev_samp(ss_ext_tax), stddev_samp(ss_item_sk), stddev_samp(ss_sales_price), stddev_samp(ss_addr_sk) from store_sales where limit 100").show}
+-----------------+-----------------------+---------------------------------------+---------------------------+---------------------------------------+
|count(ss_ext_tax)|stddev_samp(ss_ext_tax)|stddev_samp(CAST(ss_item_sk AS DOUBLE))|stddev_samp(ss_sales_price)|stddev_samp(CAST(ss_addr_sk AS DOUBLE))|
+-----------------+-----------------------+---------------------------------------+---------------------------+---------------------------------------+
|        275046633|      123.9019130788826|                      58892.23320053526|         34.425165741895285|                     293612.42312176584|
+-----------------+-----------------------+---------------------------------------+---------------------------+---------------------------------------+

Elapsed time: 12917ms

**Comment by rui-mo** [comment]
#### Test Result
#### row:
scala> time{spark.sql("select ss_item_sk, stddev_samp(ss_quantity) as store_sales_quantitystdev, stddev_samp(ss_list_price), stddev_samp(ss_ticket_number) from store_sales group by ss_item_sk order by ss_item_sk  limit 100").show}
+----------+-------------------------+--------------------------+---------------------------------------------+
|ss_item_sk|store_sales_quantitystdev|stddev_samp(ss_list_price)|stddev_samp(CAST(ss_ticket_number AS DOUBLE))|
+----------+-------------------------+--------------------------+---------------------------------------------+
|         1|       29.098844757713387|        45.684987487501004|                           6954564.2392293895|
|         2|       28.913715979808796|         45.67895110737274|                            6878413.014046009|
|         3|       29.245580021306424|          45.5064186915396|                            6982634.488898209|
|         4|        28.78873367452031|         46.75743482223979|                            6961028.661100444|
|         5|        29.10522946982835|        46.836166737158784|                            7075129.993478093|
|         6|       28.423247205783728|         47.15163292278171|                            7077872.927363345|
|         7|       28.986213962525092|        46.021991567234764|                            6891546.030064147|
|         8|        29.18407634779071|         46.84541619046964|                            6879552.527811598|
|         9|       28.828001208893188|        45.775777897290624|                            6960318.322029455|
|        10|       28.708234657077814|         45.16074014077233|                            6935612.930091468|
|        11|       28.669644445193697|         44.18214018142708|                            6946935.331484145|
|        12|       29.822617768430128|         45.00871667295131|                            7003550.311610291|
|        13|       28.521287693023258|         46.14608806648435|                            6917378.062627748|
|        14|       29.014408610915886|         46.36279201815749|                            6980381.787024499|
|        15|        28.42171789412903|         46.74001626767024|                            6818545.977606066|
|        16|       28.630078175297328|           45.937345156158|                            6998882.089007459|
|        17|       29.554911275819062|        46.308432282054504|                            6954240.964916541|
|        18|       28.762456466837115|         45.44823887552419|                            7114689.786164044|
|        19|       28.872599229536743|        45.913689187625245|                            7093953.951919404|
|        20|       29.376501344444247|         45.72323578704782|                            6878579.662488168|
+----------+-------------------------+--------------------------+---------------------------------------------+
only showing top 20 rows

Elapsed time: 17556ms


scala> time{spark.sql("select count(ss_ext_tax), stddev_samp(ss_ext_tax), stddev_samp(ss_item_sk), stddev_samp(ss_sales_price), stddev_samp(ss_addr_sk) from store_sales where limit 100").show}
+-----------------+-----------------------+---------------------------------------+---------------------------+---------------------------------------+
|count(ss_ext_tax)|stddev_samp(ss_ext_tax)|stddev_samp(CAST(ss_item_sk AS DOUBLE))|stddev_samp(ss_sales_price)|stddev_samp(CAST(ss_addr_sk AS DOUBLE))|
+-----------------+-----------------------+---------------------------------------+---------------------------+---------------------------------------+
|        275046633|     126.56286593257153|                      58892.23320053515|         34.630854050151605|                      288644.8411860637|
+-----------------+-----------------------+---------------------------------------+---------------------------+---------------------------------------+

Elapsed time: 7845ms

#### col:
groupby no codegen:
scala> time{spark.sql("select ss_item_sk, stddev_samp(ss_quantity) as store_sales_quantitystdev, stddev_samp(ss_list_price), stddev_samp(ss_ticket_number) from store_sales group by ss_item_sk order by ss_item_sk  limit 100").show}
+----------+-------------------------+--------------------------+---------------------------------------------+
|ss_item_sk|store_sales_quantitystdev|stddev_samp(ss_list_price)|stddev_samp(CAST(ss_ticket_number AS DOUBLE))|
+----------+-------------------------+--------------------------+---------------------------------------------+
|         1|       29.098844757713387|        45.684987487501026|                            6954564.239229389|
|         2|       28.913715979808796|         45.67895110737276|                             6878413.01404601|
|         3|       29.245580021306413|          45.5064186915396|                             6982634.48889821|
|         4|        28.78873367452032|        46.757434822239794|                            6961028.661100443|
|         5|        29.10522946982835|          46.8361667371588|                            7075129.993478092|
|         6|       28.423247205783735|        47.151632922781715|                            7077872.927363345|
|         7|        28.98621396252509|         46.02199156723476|                            6891546.030064147|
|         8|         29.1840763477907|         46.84541619046964|                            6879552.527811596|
|         9|        28.82800120889318|        45.775777897290624|                            6960318.322029454|
|        10|       28.708234657077796|         45.16074014077233|                            6935612.930091465|
|        11|       28.669644445193697|         44.18214018142707|                            6946935.331484145|
|        12|       29.822617768430124|        45.008716672951294|                             7003550.31161029|
|        13|        28.52128769302326|         46.14608806648437|                            6917378.062627748|
|        14|         29.0144086109159|         46.36279201815749|                            6980381.787024497|
|        15|       28.421717894129024|         46.74001626767024|                            6818545.977606066|
|        16|       28.630078175297324|         45.93734515615802|                           6998882.0890074605|
|        17|       29.554911275819055|         46.30843228205449|                            6954240.964916546|
|        18|        28.76245646683711|          45.4482388755242|                            7114689.786164045|
|        19|        28.87259922953674|        45.913689187625245|                            7093953.951919402|
|        20|        29.37650134444424|        45.723235787047805|                            6878579.662488163|
+----------+-------------------------+--------------------------+---------------------------------------------+
only showing top 20 rows

Elapsed time: 26994ms

groupby with codegen:

scala> time{spark.sql("select ss_item_sk, stddev_samp(ss_quantity) as store_sales_quantitystdev, stddev_samp(ss_list_price), stddev_samp(ss_ticket_number) from store_sales group by ss_item_sk order by ss_item_sk  limit 100").show}
+----------+-------------------------+--------------------------+---------------------------------------------+
|ss_item_sk|store_sales_quantitystdev|stddev_samp(ss_list_price)|stddev_samp(CAST(ss_ticket_number AS DOUBLE))|
+----------+-------------------------+--------------------------+---------------------------------------------+
|         1|       31.069967002932536|         48.41008394114372|                           6954564.2392293895|
|         2|        30.66174886272615|        47.681808381512276|                            6878413.014046009|
|         3|       30.442250662956965|         47.46347255493236|                            6982634.488898207|
|         4|         30.0407470036144|         48.58808088454524|                            6961028.661100446|
|         5|       30.552962301935743|         49.38439941736051|                            7075129.993478092|
|         6|        29.52804473996775|         48.42405043803398|                            7077872.927363348|
|         7|         31.0816842324423|         48.46162877881877|                            6891546.030064146|
|         8|        30.89443293725506|        49.629648585910175|                              6879552.5278116|
|         9|       30.089147228069034|        47.418590671398285|                            6960318.322029453|
|        10|       30.210749395258638|         48.17449581015934|                            6935612.930091466|
|        11|       29.795580619505337|         46.40456600293818|                           6946935.3314841455|
|        12|       30.859915634322064|         47.02131505037968|                            7003550.311610292|
|        13|        30.12285582637442|        48.368563541313954|                            6917378.062627747|
|        14|       30.789936112923954|         48.58091914984182|                              6980381.7870245|
|        15|       29.985446117097954|         48.80980124952307|                            6818545.977606066|
|        16|       30.174576065123194|         47.87510805267051|                            6998882.089007459|
|        17|       30.456853262265188|          48.1188986563067|                            6954240.964916545|
|        18|       29.946963599701363|         47.53902976938453|                            7114689.786164045|
|        19|       30.744165009957044|        48.267782260722264|                            7093953.951919405|
|        20|       31.002865951274533|        47.994500842237144|                            6878579.662488164|
+----------+-------------------------+--------------------------+---------------------------------------------+
only showing top 20 rows

Elapsed time: 27169ms

scala> time{spark.sql("select count(ss_ext_tax), stddev_samp(ss_ext_tax), stddev_samp(ss_item_sk), stddev_samp(ss_sales_price), stddev_samp(ss_addr_sk) from store_sales where limit 100").show}
+-----------------+-----------------------+---------------------------------------+---------------------------+---------------------------------------+
|count(ss_ext_tax)|stddev_samp(ss_ext_tax)|stddev_samp(CAST(ss_item_sk AS DOUBLE))|stddev_samp(ss_sales_price)|stddev_samp(CAST(ss_addr_sk AS DOUBLE))|
+-----------------+-----------------------+---------------------------------------+---------------------------+---------------------------------------+
|        275046633|      123.9019130788826|                      58892.23320053526|         34.425165741895285|                     293612.42312176584|
+-----------------+-----------------------+---------------------------------------+---------------------------+---------------------------------------+

Elapsed time: 12917ms

**Comment by rui-mo** [comment]
The test result shows that some cols are nearly the same as row result, but some are slightly different. My speculation is that we replaced null double to 0.0 due to NullPointerException, but null should be ignored when calculating stddev. Will this have an effect?      
case t: DoubleType =>
        value match {
          case null =>
            (TreeBuilder.makeLiteral(0.0: java.lang.Double), resultType)

**Comment by rui-mo** [comment]
@xuechendi @zhouyuan Could you help review this patch? The test results are listed above, thanks.

File: oap-native-sql/core/src/main/java/com/intel/oap/vectorized/ArrowWritableColumnVector.java [hunk 1]
```diff
@@ -960,6 +960,7 @@ final UTF8String getUTF8String(int rowId) {
 
     @Override
     final double getDouble(int rowId) {
+      if (isNullAt(rowId)) return 0.0;
```
Comment (rui-mo) [inline oap-native-sql/core/src/main/java/com/intel/oap/vectorized/ArrowWritableColumnVector.java:963]: to bypass "the value at index is null", is this reasonable?

File: oap-native-sql/cpp/src/codegen/arrow_compute/ext/typed_action_codegen_impl.h [hunk 2]
```diff
@@ -59,6 +59,9 @@ class TypedActionCodeGenImpl {
     RETURN_NOT_OK(MakeCodeGenRegister(func_node, &node_tmp));
     signature_ss << std::hex << std::hash<std::string>{}(node_tmp->GetFingerprint());
     auto name = "projection_" + signature_ss.str();
+    for (auto index : input_index_list_) {
+      name += "_" + std::to_string(index);
+    }
```
Comment (rui-mo) [inline oap-native-sql/cpp/src/codegen/arrow_compute/ext/typed_action_codegen_impl.h:64]: a fix for duplicated projector name

**Comment by rui-mo** [comment]
nullable = false for partial stage result n, avg and m2 
nullable = true for final stage result
corner case:
only one non-null row: stddev_samp = NaN
all null value: stddev_samp =null

**Comment by rui-mo** [comment]
@xuechendi Verified tpch queries on local env. If no other issue, we can merge it.

**Comment by xuechendi** [comment]
@rui-mo , which tpcds queries use aggr stddev? I am thinking to merge this PR after we verified on those queries, please list here, I'll try to fix remaining issue of those queries to get correct result.

**Comment by rui-mo** [comment]
stddev_samp related queries: q17, q39a, q39b
enable codegen.hashAggregate: q17 failed in join, q39a and q39b are runnable (different result)
disable codegen.hashAggregate: q17 failed in conditionProject, q39a and q39b are runnable (correct result)

**Comment by xuechendi** [comment]
@rui-mo , I see, I'll check with these queries

**Comment by xuechendi** [comment]
@rui-mo , BTW, according to your summary, do you think you can split this PR into two, I think we can merge wo codegen hashaggr firstly

**Pull request closed without merging**